src/libmongoc/tests/mock_server/mock-server.c

```diff
@@ -1218,12 +1223,14 @@ mock_server_receives_kill_cursors (mock_server_t *server, int64_t cursor_id)
 void
 mock_server_hangs_up (request_t *request)
 {
+   reply_t *reply;
    test_suite_mock_server_log ("%5.2f  %hu <- %hu \thang up!",
                                mock_server_get_uptime_sec (request->server),
                                request->client_port,
                                request_get_server_port (request));
-
-   mongoc_stream_close (request->client);
+   reply = bson_malloc0 (sizeof (reply_t));
+   reply->type = HANGUP;
+   q_put (request->replies, reply);
```

Comment:
The mock server spawns a worker thread for every connection. And each worker thread maintains a thread-safe queue for sending replies back.

But the behavior of `mock_server_hangs_up` was to close the server side stream from the main thread. TSan flags a data race because in the process of closing, it overwrites the stream's file descriptor with an invalid one, causing a write/read race.